<H2> EXPLORATORY DATA ANALYSIS <H2>

In this project we will import the already cleaned data set "Automobiles.csv" and look at different factors that affect the price of a car

First we import the required libraries

In [1]:
import pandas as pd
import numpy as np

Let's import the data set

In [2]:
df=pd.read_csv('automobile.csv')
df.head()

symboling  normalized-losses         make fuel-type aspiration  \
0        3.0          99.542289  alfa-romero       gas        std   
1        3.0          99.542289  alfa-romero       gas        std   
2        1.0          99.542289  alfa-romero       gas        std   
3        2.0         164.000000         audi       gas        std   
4        2.0         164.000000         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0          130         mpfi  3.47    2.68               9.0      111.0   
1          130         mpfi  3.47    2.68               9.0      111.0   
2          152         mpfi  2.68    3.47               9.0      154.0   
3          109         mpfi  3.19    3.40              10.0      102.0   
4          136         mpfi  3.19    3.40               8.0      115.0   

   peak-rpm city-mpg  highway-mpg  price  
0    5000.0       21           27  13495  
1    5000.0       21           27  16500  
2    5000.0       19           26  16500  
3    5500.0       24           30  13950  
4    5500.0       18           22  17450  

[5 rows x 26 columns]

Firstly, there are two types of values in the dataset:

1. Continuous Numerical Values: Continuous numerical variables are variables that may contain any value within some range. They can be of type "int64" or "float64". A great way to visualize these variables is by using scatterplots with fitted lines. e.g Length,Width,etc.

2. Categorical Values:These are variables that describe a 'characteristic' of a data unit, and are selected from a small group of categories. The categorical variables can have the type "object" or "int64". A good way to visualize categorical variables is by using boxplots. E.g Body styles, Drive wheels, etc

<h3> Descriptive Statistical Analysis </h3>

Let's first take a look at the variables by utilizing a description method.

The describe function automatically computes basic statistics for all continuous variables. Any NaN values are automatically skipped in these statistics.


This will show:

    the count of that variable
    the mean
    the standard deviation (std)
    the minimum value
    the IQR (Interquartile Range: 25%, 50% and 75%)
    the maximum value

We can apply the method "describe" as follows:

In [3]:
df.describe()

symboling  normalized-losses  wheel-base      length       width  \
count  130.000000         130.000000  130.000000  130.000000  130.000000   
mean     1.184615         120.917719   97.740000  172.173846   65.724615   
std      1.345591          34.880566    5.670878   12.459339    2.130984   
min     -2.000000          65.000000   86.600000  141.100000   60.300000   
25%      1.000000          99.542289   93.700000  165.300000   64.000000   
50%      1.000000         111.500000   96.300000  171.700000   65.500000   
75%      2.000000         144.250000  100.850000  178.250000   66.500000   
max      3.000000         256.000000  115.600000  202.600000   72.000000   

           height  curb-weight  engine-size        bore      stroke  \
count  130.000000   130.000000   130.000000  130.000000  130.000000   
mean    53.395385  2483.661538   124.461538    3.295385    3.270385   
std      2.391421   493.763235    36.016825    0.283678    0.268566   
min     48.800000  1488.000000    61.000000    2.540000    2.070000   
25%     51.400000  2025.000000    97.000000    3.035000    3.150000   
50%     53.700000  2399.000000   121.000000    3.240000    3.290000   
75%     55.500000  2841.250000   141.000000    3.570000    3.407500   
max     59.800000  3770.000000   304.000000    3.940000    3.900000   

       compression-ratio  horsepower     peak-rpm    city-mpg  highway-mpg  \
count         130.000000  130.000000   130.000000  130.000000   130.000000   
mean            9.827077  103.776923  5167.307692   25.692308    31.292308   
std             3.489471   37.296271   438.757563    6.572017     6.870278   
min             7.000000   48.000000  4200.000000   14.000000    16.000000   
25%             8.700000   69.000000  4825.000000   20.000000    25.250000   
50%             9.200000  100.500000  5200.000000   24.000000    30.000000   
75%             9.400000  116.000000  5500.000000   31.000000    37.000000   
max            23.000000  207.000000  6600.000000   49.000000    54.000000   

              price  
count    130.000000  
mean   12679.146154  
std     7596.033089  
min     5118.000000  
25%     7360.500000  
50%     9969.500000  
75%    16482.500000  
max    45400.000000

The default setting of "describe" skips variables of type object. We can apply the method "describe" on the variables of type 'object' as follows:

In [4]:
df.describe(include=['object'])

make fuel-type aspiration num-of-doors body-style drive-wheels  \
count      130       130        130          130        130          130   
unique      19         2          2            2          5            3   
top     toyota       gas        std          two  hatchback          fwd   
freq        22       121        105           75         56           82   

       engine-location engine-type num-of-cylinders fuel-system  
count              130         130              130         130  
unique               2           5                5           7  
top              front         ohc             four        mpfi  
freq               127         105              104          61

<h3>Correlation and Causation</h3>

Correlation: a measure of the extent of interdependence between variables

Causation: the relationship between cause and effect between two variables.

It is important to know the difference between these two. Correlation does not imply causation. Determining correlation is much simpler  the determining causation as causation may require independent experimentation.


Pearson Correlation
The Pearson Correlation measures the linear dependence between two variables X and Y.
The resulting coefficient is a value between -1 and 1 inclusive, where: 
    
     1: Perfect positive linear correlation.
     0: No linear correlation, the two variables most likely do not affect each other.
    -1: Perfect negative linear correlation.


Pearson Correlation is the default method of the function "corr". Like before, we can calculate the Pearson Correlation of the of the 'int64' or 'float64' variables.


In [5]:
df.corr()

symboling  normalized-losses  wheel-base    length  \
symboling           1.000000           0.448870   -0.610307 -0.343862   
normalized-losses   0.448870           1.000000   -0.256944 -0.124925   
wheel-base         -0.610307          -0.256944    1.000000  0.856055   
length             -0.343862          -0.124925    0.856055  1.000000   
width              -0.269238          -0.065097    0.807593  0.852851   
height             -0.532610          -0.461943    0.636788  0.509726   
curb-weight        -0.197016          -0.084135    0.724754  0.864522   
engine-size        -0.017607          -0.037327    0.522259  0.676878   
bore               -0.079796          -0.178750    0.417694  0.571033   
stroke             -0.088576          -0.000935    0.185374  0.105724   
compression-ratio  -0.192387          -0.169986    0.246721  0.148434   
horsepower          0.089645           0.061079    0.340207  0.589817   
peak-rpm            0.196095           0.377596   -0.312568 -0.236418   
city-mpg           -0.025961          -0.040643   -0.464441 -0.696025   
highway-mpg         0.051138          -0.008882   -0.538534 -0.721002   
price              -0.052705          -0.068683    0.521231  0.671473   

                      width    height  curb-weight  engine-size      bore  \
symboling         -0.269238 -0.532610    -0.197016    -0.017607 -0.079796   
normalized-losses -0.065097 -0.461943    -0.084135    -0.037327 -0.178750   
wheel-base         0.807593  0.636788     0.724754     0.522259  0.417694   
length             0.852851  0.509726     0.864522     0.676878  0.571033   
width              1.000000  0.345736     0.877501     0.755730  0.493405   
height             0.345736  1.000000     0.278911     0.071378  0.182250   
curb-weight        0.877501  0.278911     1.000000     0.882153  0.654855   
engine-size        0.755730  0.071378     0.882153     1.000000  0.632904   
bore               0.493405  0.182250     0.654855     0.632904  1.000000   
stroke             0.202996 -0.106743     0.134056     0.127646 -0.010243   
compression-ratio  0.172179  0.304843     0.125844     0.004201 -0.109344   
horsepower         0.607904 -0.032570     0.796221     0.823046  0.627208   
peak-rpm          -0.226305 -0.253703    -0.225578    -0.222187 -0.172372   
city-mpg          -0.660687 -0.083744    -0.802686    -0.743270 -0.630374   
highway-mpg       -0.711597 -0.121412    -0.844821    -0.776948 -0.627549   
price              0.754461  0.191482     0.846365     0.889200  0.563228   

                     stroke  compression-ratio  horsepower  peak-rpm  \
symboling         -0.088576          -0.192387    0.089645  0.196095   
normalized-losses -0.000935          -0.169986    0.061079  0.377596   
wheel-base         0.185374           0.246721    0.340207 -0.312568   
length             0.105724           0.148434    0.589817 -0.236418   
width              0.202996           0.172179    0.607904 -0.226305   
height            -0.106743           0.304843   -0.032570 -0.253703   
curb-weight        0.134056           0.125844    0.796221 -0.225578   
engine-size        0.127646           0.004201    0.823046 -0.222187   
bore              -0.010243          -0.109344    0.627208 -0.172372   
stroke             1.000000           0.131929    0.046165 -0.187054   
compression-ratio  0.131929           1.000000   -0.198528 -0.315876   
horsepower         0.046165          -0.198528    1.000000  0.103935   
peak-rpm          -0.187054          -0.315876    0.103935  1.000000   
city-mpg          -0.035420           0.301062   -0.848116 -0.057686   
highway-mpg       -0.057577           0.254556   -0.836122 -0.013135   
price             -0.051589           0.103733    0.797357 -0.068482   

                   city-mpg  highway-mpg     price  
symboling         -0.025961     0.051138 -0.052705  
normalized-losses -0.040643    -0.008882 -0.068683  
wheel-base        -0.464441    -0.538534  0.521231  
length            -0.696025   

Sometimes we would like to know the significant of the correlation estimate.

P-value

What is this P-value? The P-value is the probability value that the correlation between these two variables is statistically significant. Normally, we choose a significance level of 0.05, which means that we are 95% confident that the correlation between the variables is significant.

By convention, when the

    p-value is <0.001: we say there is strong evidence that the correlation is significant.
    the p-value is <0.05: there is moderate evidence that the correlation is significant.
    the p-value is <0.1: there is weak evidence that the correlation is significant.
    the p-value is >0.1: there is no evidence that the correlation is significant.

We can obtain this information using "stats" module in the "scipy" library.

In [7]:
! pip install scipy
from scipy import stats


You should consider upgrading via the 'C:\Python\Python39\python.exe -m pip install --upgrade pip' command.


Wheel-Base vs. Price

Let's calculate the Pearson Correlation Coefficient and P-value of 'wheel-base' and 'price'.


In [8]:
pearson_coef,p_value=stats.pearsonr(df['wheel-base'],df['price'])
print("The Pearson Correlation Coefficient is", pearson_coef, " with a P-value of P = ", p_value)  

The Pearson Correlation Coefficient is 0.5212311698596392  with a P-value of P =  2.0439212035694013e-10


Conclusion:

Since the p-value is <0.001, the correlation between wheel-base and price is statistically significant, although the linear relationship isn't extremely strong (~0.585).

Length vs. Price

Let's calculate the Pearson Correlation Coefficient and P-value of 'length' and 'price'.


In [9]:
pearson_coef, p_value = stats.pearsonr(df['length'], df['price'])
print("The Pearson Correlation Coefficient is", pearson_coef, " with a P-value of P = ", p_value)  

The Pearson Correlation Coefficient is 0.6714732220741576  with a P-value of P =  2.266583638104945e-18


Conclusion:

Since the p-value is <0.001, the correlation between length and price is statistically significant, and the linear relationship is moderately strong (~0.691).

Horsepower vs. Price

Let's calculate the Pearson Correlation Coefficient and P-value of 'horsepower' and 'price'.


In [10]:
pearson_coef, p_value = stats.pearsonr(df['horsepower'], df['price'])
print("The Pearson Correlation Coefficient is", pearson_coef, " with a P-value of P = ", p_value)  

The Pearson Correlation Coefficient is 0.7973566903717995  with a P-value of P =  7.437027405763391e-30


Conclusion:

Since the p-value is <0.001, the correlation between horsepower and price is statistically significant, and the linear relationship is quite strong (~0.809, close to 1).

Width vs. Price

Let's calculate the Pearson Correlation Coefficient and P-value of 'width' and 'price':


In [11]:
pearson_coef, p_value = stats.pearsonr(df['width'], df['price'])
print("The Pearson Correlation Coefficient is", pearson_coef, " with a P-value of P =", p_value ) 

The Pearson Correlation Coefficient is 0.7544609938999314  with a P-value of P = 3.633008353695635e-25


Conclusion:

Since the p-value is < 0.001, the correlation between width and price is statistically significant, and the linear relationship is quite strong (~0.751).


Curb-Weight vs. Price

Let's calculate the Pearson Correlation Coefficient and P-value of 'curb-weight' and 'price':


In [12]:
pearson_coef, p_value = stats.pearsonr(df['curb-weight'], df['price'])
print( "The Pearson Correlation Coefficient is", pearson_coef, " with a P-value of P = ", p_value)  

The Pearson Correlation Coefficient is 0.8463652318717515  with a P-value of P =  7.911735042406564e-37


Conclusion:

Since the p-value is <0.001, the correlation between curb-weight and price is statistically significant, and the linear relationship is quite strong (~0.834).


Engine-Size vs. Price

Let's calculate the Pearson Correlation Coefficient and P-value of 'engine-size' and 'price':


In [13]:
pearson_coef, p_value = stats.pearsonr(df['engine-size'], df['price'])
print("The Pearson Correlation Coefficient is", pearson_coef, " with a P-value of P =", p_value) 

The Pearson Correlation Coefficient is 0.8892001115164775  with a P-value of P = 2.690999366963704e-45


Conclusion:

Since the p-value is <0.001, the correlation between engine-size and price is statistically significant, and the linear relationship is very strong (~0.872).

Bore vs. Price

Let's calculate the Pearson Correlation Coefficient and P-value of 'bore' and 'price':


In [14]:
pearson_coef, p_value = stats.pearsonr(df['bore'], df['price'])
print("The Pearson Correlation Coefficient is", pearson_coef, " with a P-value of P =  ", p_value ) 

The Pearson Correlation Coefficient is 0.5632282630129494  with a P-value of P =   3.0451365719945618e-12


Conclusion:

Since the p-value is <0.001, the correlation between bore and price is statistically significant, but the linear relationship is only moderate (~0.521).

We can relate the process for each 'city-mpg' and 'highway-mpg':


City-mpg vs. Price

In [15]:
pearson_coef, p_value = stats.pearsonr(df['city-mpg'], df['price'])
print("The Pearson Correlation Coefficient is", pearson_coef, " with a P-value of P = ", p_value)  

The Pearson Correlation Coefficient is -0.7064618349061609  with a P-value of P =  6.023105087635934e-21


Conclusion:

Since the p-value is <0.001, the correlation between city-mpg and price is statistically significant, and the coefficient of about -0.687 shows that the relationship is negative and moderately strong.

Highway-mpg vs. Price


In [16]:
pearson_coef, p_value = stats.pearsonr(df['highway-mpg'], df['price'])
print( "The Pearson Correlation Coefficient is", pearson_coef, " with a P-value of P = ", p_value ) 

The Pearson Correlation Coefficient is -0.7339950747419518  with a P-value of P =  2.957881570195919e-23


Conclusion:

Since the p-value is < 0.001, the correlation between highway-mpg and price is statistically significant, and the coefficient of about -0.705 shows that the relationship is negative and moderately strong.


<h3>ANOVA: Analysis of Variance</h3>

The Analysis of Variance (ANOVA) is a statistical method used to test whether there are significant differences between the means of two or more groups. ANOVA returns two parameters:

F-test score: ANOVA assumes the means of all groups are the same, calculates how much the actual means deviate from the assumption, and reports it as the F-test score. A larger score means there is a larger difference between the means.

P-value: P-value tells how statistically significant our calculated score value is.

If our price variable is strongly correlated with the variable we are analyzing, we expect ANOVA to return a sizeable F-test score and a small p-value.


Drive Wheels

Since ANOVA analyzes the difference between different groups of the same variable, the groupby function will come in handy. Because the ANOVA algorithm averages the data automatically, we do not need to take the average before hand.

To see if different types of 'drive-wheels' impact 'price', we group the data.

First we create a data frame from the dataset containing the promising categorical values as 'df_gptest'.


In [17]:
df_gptest = df[['drive-wheels','engine-location','engine-type','num-of-cylinders','fuel-system','body-style','price']]
grouped_test2=df_gptest[['drive-wheels', 'price']].groupby(['drive-wheels'])
grouped_test2.head(2)

drive-wheels  price
0           rwd  13495
1           rwd  16500
3           fwd  13950
4           4wd  17450
5           fwd  15250
87          4wd   7603

In [18]:
f_val, p_val = stats.f_oneway(grouped_test2.get_group('fwd')['price'], grouped_test2.get_group('rwd')['price'], grouped_test2.get_group('4wd')['price'])  
 
print( "ANOVA results: F=", f_val, ", P =", p_val)   

ANOVA results: F= 40.251291357052544 , P = 2.8876437412053394e-14


This is a great result with a large F-test score showing a strong correlation and a P-value of almost 0 implying almost certain statistical significance. But does this mean all three tested groups are all this highly correlated?

Let's examine them separately.


FWD vs RWD

In [19]:
f_val, p_val = stats.f_oneway(grouped_test2.get_group('fwd')['price'], grouped_test2.get_group('rwd')['price'])  
 
print( "ANOVA results: F=", f_val, ", P =", p_val )

ANOVA results: F= 80.7265083140164 , P = 3.186566267213619e-15


4WD vs RWD

In [20]:
f_val, p_val = stats.f_oneway(grouped_test2.get_group('4wd')['price'], grouped_test2.get_group('rwd')['price'])  
   
print( "ANOVA results: F=", f_val, ", P =", p_val)  

ANOVA results: F= 1.0847780182854088 , P = 0.30307441475273617


4WD vs FWD

In [21]:
f_val, p_val = stats.f_oneway(grouped_test2.get_group('4wd')['price'], grouped_test2.get_group('fwd')['price'])  
 
print("ANOVA results: F=", f_val, ", P =", p_val)   

ANOVA results: F= 1.5789573224978852 , P = 0.21247882348856417


We can see that only the FWD vs RWD shows a strong correlation as well as statistical significance.

Now, let's try for the other categorical values.

Engine-location

In [22]:
grouped_test2=df_gptest[['engine-location', 'price']].groupby(['engine-location'])
grouped_test2.head(2)

engine-location  price
0            front  13495
1            front  16500
76            rear  32528
77            rear  34028

In [23]:
f_val, p_val = stats.f_oneway(grouped_test2.get_group('front')['price'], grouped_test2.get_group('rear')['price'])  
 
print( "ANOVA results: F=", f_val, ", P =", p_val)  

ANOVA results: F= 31.392210037322467 , P = 1.2335318028697013e-07


Conclusion: A large F-test score shows a strong correlation and a P-value of almost 0 means there is statistical significance.

Engine type

In [24]:
grouped_test2=df_gptest[['engine-type', 'price']].groupby(['engine-type'])
grouped_test2.head(2)

engine-type  price
0         dohc  13495
1         dohc  16500
2         ohcv  16500
3          ohc  13950
4          ohc  17450
11           l   5151
38        ohcv  34184
76        ohcf  32528
77        ohcf  34028

In [25]:
f_val, p_val = stats.f_oneway(grouped_test2.get_group('dohc')['price'], grouped_test2.get_group('ohcv')['price'], grouped_test2.get_group('ohc')['price'], grouped_test2.get_group('l')['price'], grouped_test2.get_group('ohcf')['price'])  
 
print( "ANOVA results: F=", f_val, ", P =", p_val) 

ANOVA results: F= 12.867632233147496 , P = 8.391203012529898e-09


Conclusion: Although a small P-value, there isn't much correlation due to a small F-score.

Number of Cylinders

In [26]:
grouped_test2=df_gptest[['num-of-cylinders', 'price']].groupby(['num-of-cylinders'])
grouped_test2.head(2)

num-of-cylinders  price
0              four  13495
1              four  16500
2               six  16500
4              five  17450
5              five  15250
10              six  24565
11            three   5151
38            eight  34184
39            eight  35056

In [27]:
f_val, p_val = stats.f_oneway(grouped_test2.get_group('three')['price'], grouped_test2.get_group('four')['price'], grouped_test2.get_group('five')['price'], grouped_test2.get_group('six')['price'], grouped_test2.get_group('eight')['price'])  
 
print( "ANOVA results: F=", f_val, ", P =", p_val)   

ANOVA results: F= 53.766708831915295 , P = 2.765974813131859e-26


Conclusion: A large F-test score shows a strong correlation and a P-value of almost 0 means there is statistical significance.

Fuel-system

In [28]:
grouped_test2=df_gptest[['fuel-system', 'price']].groupby(['fuel-system'])
grouped_test2.head(2)

fuel-system  price
0         mpfi  13495
1         mpfi  16500
11        2bbl   5151
12        2bbl   6295
20         mfi  12964
21        1bbl   6479
22        1bbl   6855
27        spfi  11048
35         idi  25552
36         idi  28248
45        spdi   7689
46        spdi   9959

In [29]:
f_val, p_val = stats.f_oneway(grouped_test2.get_group('mpfi')['price'], grouped_test2.get_group('2bbl')['price'], grouped_test2.get_group('mfi')['price'], grouped_test2.get_group('1bbl')['price'], grouped_test2.get_group('spfi')['price'],grouped_test2.get_group('idi')['price'],grouped_test2.get_group('spdi')['price'])  
 
print( "ANOVA results: F=", f_val, ", P =", p_val) 

ANOVA results: F= 12.65319808729073 , P = 4.4057464475211666e-11


Conclusion: Although a small P-value, there isn't much correlation due to a small F-score.

Body Style

In [30]:
grouped_test2=df_gptest[['body-style', 'price']].groupby(['body-style'])
grouped_test2.head(2)

body-style  price
0   convertible  13495
1   convertible  16500
2     hatchback  16500
3         sedan  13950
4         sedan  17450
7         wagon  18920
11    hatchback   5151
19        wagon   8921
40      hardtop  45400
64      hardtop   8249

In [31]:
f_val, p_val = stats.f_oneway(grouped_test2.get_group('convertible')['price'], grouped_test2.get_group('hatchback')['price'], grouped_test2.get_group('sedan')['price'], grouped_test2.get_group('wagon')['price'], grouped_test2.get_group('hardtop')['price'])  
 
print( "ANOVA results: F=", f_val, ", P =", p_val) 

ANOVA results: F= 7.949421024701909 , P = 9.63971484925619e-06


Conclusion: Although a small P-value, there isn't much correlation due to a small F-score.

<h2>Conclusion: Important Variables</h2>

We now have a better idea of what our data looks like and which variables are important to take into account when predicting the car price. We have narrowed it down to the following variables:

Continuous numerical variables:

    Length
    Width
    Curb-weight
    Engine-size
    Horsepower
    City-mpg
    Highway-mpg
    Wheel-base
    Bore

Categorical variables:

    Drive-wheels
    Num-of-cylinders
    Engine-location